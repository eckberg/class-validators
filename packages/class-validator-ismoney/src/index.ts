import {
  registerDecorator,
  ValidationArguments,
  ValidationOptions,
  ValidatorConstraint,
  ValidatorConstraintInterface,
} from 'class-validator';

import codes from './codes';
import { IMoney } from './Money';

const settingsDefault = {
  allowNegative: false,
  min: Number.MIN_SAFE_INTEGER,
  max: Number.MAX_SAFE_INTEGER,
  currencies: undefined,
};

interface IMoneyValidationOptions {
  allowNegative?: boolean;
  min?: number;
  max?: number;
  currencies?: string[];
}

@ValidatorConstraint({ name: 'IsMoney', async: false })
export class IsMoneyValidator implements ValidatorConstraintInterface {
  public validate(money: IMoney, args: ValidationArguments) {
    if (!money) {
      return false;
    }
    const { amount, currency } = money;
    const {
      allowNegative,
      min,
      max,
      currencies: customCurrencies,
    } = Object.assign({}, settingsDefault, args.constraints[0]);
    const isAmountCorrect =
      typeof amount === 'number' &&
      (allowNegative || amount >= 0) &&
      amount <= max &&
      amount >= min;
    const isCurrencyCorrect =
      typeof currency === 'string' && codes.includes(currency);
    const isCustomCurrencyCorrect =
      !customCurrencies ||
      (customCurrencies.length && customCurrencies.includes(currency));
    return isAmountCorrect && isCurrencyCorrect && isCustomCurrencyCorrect;
  }

  public defaultMessage(args: ValidationArguments) {
    return `(${JSON.stringify(args.value)}) is not a valid Money object`;
  }
}

export function IsMoney(
  settings?: IMoneyValidationOptions,
  validationOptions?: ValidationOptions,
) {
  return (object: object, propertyName: string) => {
    registerDecorator({
      target: object.constructor,
      propertyName,
      options: validationOptions,
      constraints: [settings],
      validator: IsMoneyValidator,
    });
  };
}
