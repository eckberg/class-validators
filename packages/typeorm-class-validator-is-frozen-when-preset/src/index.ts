import {
  registerDecorator,
  ValidationArguments,
  ValidationOptions,
  ValidatorConstraint,
  ValidatorConstraintInterface
} from 'class-validator'
import { getRepository } from 'typeorm'

interface IObject {
  [key: string]: any
}

@ValidatorConstraint({ async: true, name: 'IsFrozenWhenPreset' })
export class IsFrozenWhenPresetConstraint implements ValidatorConstraintInterface {
  public async validate(value: any, args: ValidationArguments) {
    const { targetName, property } = args

    const repository = getRepository(targetName)
    const object: IObject = args.object
    const entity: IObject | undefined = await repository.findOne(object.id)

    return !entity || !entity[property] || entity[property] === value
  }
}

export const IsFrozenWhenPreset = () => {
  const options: ValidationOptions = {
    message: 'Value is not allowed to be changed'
  }

  return (object: object, propertyName: string) => {
    registerDecorator({
      target: object.constructor,
      propertyName,
      options,
      constraints: [],
      validator: IsFrozenWhenPresetConstraint
    })
  }
}
