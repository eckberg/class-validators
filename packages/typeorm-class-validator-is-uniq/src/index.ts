import { getRepository, Not, Repository } from 'typeorm';

import {
  registerDecorator,
  ValidationArguments,
  ValidationOptions,
  ValidatorConstraint,
  ValidatorConstraintInterface,
} from 'class-validator';

export type ScopedValidationOptions = ValidationOptions & { scope?: string[] };

@ValidatorConstraint({ async: true, name: 'IsUniq' })
export class IsUniqConstraint implements ValidatorConstraintInterface {
  public async validate(value: any, args: ValidationArguments) {
    const repository = getRepository<any>(args.targetName);
    if (args.value == null) return true;
    const entity = await repository.findOne({
      where: this.buildConditions(value, args, repository),
    });

    return !entity;
  }

  private buildConditions(
    value: any,
    args: ValidationArguments,
    repository: Repository<{}>,
  ) {
    return {
      [args.property]: value,
      ...this.buildScopeConditions(args.object, args.constraints),
      ...this.buildPrimaryColumnConditions(args.object, repository),
    };
  }

  private buildScopeConditions(object: any, constraints?: string[]) {
    if (!constraints || !constraints.length) {
      return {};
    }
    return constraints.reduce(
      (acc, key) => ({
        ...acc,
        [key]: object[key],
      }),
      {},
    );
  }

  private buildPrimaryColumnConditions(
    object: any,
    repository: Repository<{}>,
  ) {
    const primaryColumnNames = repository.metadata.primaryColumns.map(
      ({ propertyName }) => propertyName,
    );

    if (!primaryColumnNames.length) {
      return {};
    }
    return primaryColumnNames.reduce((acc, name) => {
      const pkValue = object[name];
      return pkValue ? { ...acc, [name]: Not(pkValue) } : acc;
    }, {});
  }
}

/**
 * Checks if a value is uniq across all records in a database or inside a scope.
 *
 * @param validationOptions accept `scope` options and all `class-validator` options
 */
export const IsUniq = (validationOptions?: ScopedValidationOptions) => {
  return (object: object, propertyName: string) => {
    const scope = validationOptions && validationOptions.scope;
    const opts: ScopedValidationOptions = {
      message: scope
        ? `$target with $property '$value' already exists in scope: ${scope.join(
            ', ',
          )}`
        : "$target with $property '$value' already exists",
      ...validationOptions,
    };
    registerDecorator({
      target: object.constructor,
      propertyName,
      options: opts,
      constraints: scope || [],
      validator: IsUniqConstraint,
    });
  };
};
