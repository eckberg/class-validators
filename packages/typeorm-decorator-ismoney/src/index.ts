import {
  registerDecorator,
  ValidationArguments,
  ValidationOptions,
  ValidatorConstraint,
  ValidatorConstraintInterface,
} from 'class-validator';

import codes from './codes';
import { IMoney } from './Money';

const settingsDefault = {
  allowNegative: false,
  min: Number.MIN_SAFE_INTEGER,
  max: Number.MAX_SAFE_INTEGER,
  currencies: null,
};

interface IMoneyValidationOptions {
  allowNegative?: boolean;
  min?: number;
  max?: number;
  currencies?: string[];
}

@ValidatorConstraint({ name: 'IsMoney', async: false })
export class IsMoneyValidator implements ValidatorConstraintInterface {
  public validate(money: IMoney, args: ValidationArguments) {
    const { amount, currency } = money;
    const settingsOrDefault = args.constraints[0] || settingsDefault;
    const {
      allowNegative,
      min,
      max,
      currencies: customCurrencies,
    } = Object.assign({}, settingsDefault, settingsOrDefault);
    const isAmountCorrect =
      typeof amount === 'number' &&
      (allowNegative || amount >= 0) &&
      amount <= max &&
      amount >= min;
    const isCurrencyCorrect =
      typeof currency === 'string' && codes.includes(currency);
    const isCustomCurrencyCorrect =
      !customCurrencies ||
      (customCurrencies.length && customCurrencies.includes(currency));
    return isAmountCorrect && isCurrencyCorrect && isCustomCurrencyCorrect;
  }

  public defaultMessage(args: ValidationArguments) {
    // here you can provide default error message if validation failed
    return `Provided object (${JSON.stringify(
      args.value,
    )}) is too a valid Money object`;
  }
}

export function IsMoney(
  settings: IMoneyValidationOptions,
  validationOptions?: ValidationOptions,
) {
  return (object: object, propertyName: string) => {
    registerDecorator({
      target: object.constructor,
      propertyName,
      options: validationOptions,
      constraints: [settings],
      validator: IsMoneyValidator,
    });
  };
}
