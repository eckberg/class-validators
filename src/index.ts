import {
  registerDecorator,
  ValidationArguments,
  ValidationOptions,
  ValidatorConstraint,
  ValidatorConstraintInterface,
} from 'class-validator';
import { getRepository, Not, Repository } from 'typeorm';

export type ScopedValidationOptions = ValidationOptions & { scope?: string[] };

@ValidatorConstraint({ async: true, name: 'IsUniq' })
export class IsUniqConstraint implements ValidatorConstraintInterface {
  public async validate(value: any, args: ValidationArguments) {
    const repository = getRepository(args.targetName);
    const entity = await repository.findOne({
      where: this.buildConditions(value, args, repository),
    });

    return !entity;
  }

  private buildConditions(
    value: any,
    args: ValidationArguments,
    repository: Repository<{}>,
  ) {
    return {
      [args.property]: value,
      ...this.buildScopeConditions(args.object, args.constraints),
      ...this.buildPrimaryColumnConditions(args.object, repository),
    };
  }

  private buildScopeConditions(object: any, constraints?: string[]) {
    if (!constraints || !constraints.length) {
      return {};
    }
    return constraints.reduce(
      (acc, key) => ({
        ...acc,
        [key]: object[key],
      }),
      {},
    );
  }

  private buildPrimaryColumnConditions(
    object: any,
    repository: Repository<{}>,
  ) {
    const primaryColumnNames = repository.metadata.primaryColumns.map(
      ({ propertyName }) => propertyName,
    );
    if (!primaryColumnNames.length) {
      return {};
    }
    return primaryColumnNames.reduce((acc, name) => {
      const pkValue = object[name];
      return pkValue ? { acc, [name]: Not(pkValue) } : acc;
    }, {});
  }
}

export const IsUniq = (validationOptions?: ScopedValidationOptions) => {
  return (object: object, propertyName: string) => {
    const scope = validationOptions && validationOptions.scope;
    const opts: ScopedValidationOptions = {
      message: scope
        ? `$target with $property: '$value' already exists in scope: ${scope.join(
            ', ',
          )}`
        : "$target with $property: '$value' already exists",
      ...validationOptions,
    };
    registerDecorator({
      target: object.constructor,
      propertyName,
      options: opts,
      constraints: scope || [],
      validator: IsUniqConstraint,
    });
  };
};
