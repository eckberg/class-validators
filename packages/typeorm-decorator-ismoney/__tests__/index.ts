import { validate } from 'class-validator';
import { IsMoney, IsMoneyValidator } from '../src';

describe('IsMoney validation', () => {
  describe('Validator', () => {
    it('accepts valid money object', () => {
      const validator = new IsMoneyValidator();
      expect(validator.validate({ amount: 100, currency: 'EUR' })).toBeTrue();
    });

    it('rejects money object with missing amount', () => {
      const validator = new IsMoneyValidator();
      expect(validator.validate({ currency: 'EUR' })).toBeFalse();
    });

    it('rejects money object with negative amount', () => {
      const validator = new IsMoneyValidator();
      expect(validator.validate({ amount: -12, currency: 'EUR' })).toBeFalse();
    });

    it('rejects money object with non-numeric amount', () => {
      const validator = new IsMoneyValidator();
      expect(
        validator.validate({ amount: 'Plethora', currency: 'EUR' }),
      ).toBeFalse();
    });

    it('rejects money object with missing currency', () => {
      const validator = new IsMoneyValidator();
      expect(validator.validate({ amount: 1701 })).toBeFalse();
    });

    it('rejects money object with invalid currency code', () => {
      const validator = new IsMoneyValidator();
      expect(validator.validate({ amount: 1701, currency: 'FOO' })).toBeFalse();
    });
  });

  describe('Decorator', () => {
    class GuineaPig {
      @IsMoney()
      public price: any;
    }

    it('accepts valid class', () => {
      const littlePiggy = new GuineaPig();
      littlePiggy.price = {
        amount: 42,
        currency: 'CHF',
      };
      expect.assertions(1);
      validate(littlePiggy).then(errors => {
        expect(errors).toBeEmpty();
      });
    });

    it('rejects invalid class', () => {
      const littlePiggy = new GuineaPig();
      littlePiggy.price = {
        amount: 'Enough',
        currency: 'Gold',
      };
      expect.assertions(1);
      validate(littlePiggy).then(errors => {
        expect(errors).toMatchSnapshot();
      });
    });
  });
});
