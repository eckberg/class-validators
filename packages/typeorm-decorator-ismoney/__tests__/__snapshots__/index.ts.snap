// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IsMoney validation checks max 1`] = `
Array [
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "IsMoney": "Provided object ({\\"amount\\":2400,\\"currency\\":\\"EUR\\"}) is too a valid Money object",
    },
    "property": "price",
    "target": GuineaPig {
      "price": Object {
        "amount": 2400,
        "currency": "EUR",
      },
    },
    "value": Object {
      "amount": 2400,
      "currency": "EUR",
    },
  },
]
`;

exports[`IsMoney validation checks min 1`] = `
Array [
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "IsMoney": "Provided object ({\\"amount\\":-2000,\\"currency\\":\\"USD\\"}) is too a valid Money object",
    },
    "property": "price",
    "target": CustomGuineaPig {
      "price": Object {
        "amount": -2000,
        "currency": "USD",
      },
    },
    "value": Object {
      "amount": -2000,
      "currency": "USD",
    },
  },
]
`;

exports[`IsMoney validation rejects currency which is not in the custom list 1`] = `
Array [
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "IsMoney": "Provided object ({\\"amount\\":100,\\"currency\\":\\"BYN\\"}) is too a valid Money object",
    },
    "property": "price",
    "target": CustomGuineaPig {
      "price": Object {
        "amount": 100,
        "currency": "BYN",
      },
    },
    "value": Object {
      "amount": 100,
      "currency": "BYN",
    },
  },
]
`;

exports[`IsMoney validation rejects invalid class 1`] = `
Array [
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "IsMoney": "Provided object ({\\"amount\\":\\"Enough\\",\\"currency\\":\\"Gold\\"}) is too a valid Money object",
    },
    "property": "price",
    "target": GuineaPig {
      "price": Object {
        "amount": "Enough",
        "currency": "Gold",
      },
    },
    "value": Object {
      "amount": "Enough",
      "currency": "Gold",
    },
  },
]
`;

exports[`IsMoney validation rejects negative by default 1`] = `
Array [
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "IsMoney": "Provided object ({\\"amount\\":-2400,\\"currency\\":\\"EUR\\"}) is too a valid Money object",
    },
    "property": "price",
    "target": GuineaPig {
      "price": Object {
        "amount": -2400,
        "currency": "EUR",
      },
    },
    "value": Object {
      "amount": -2400,
      "currency": "EUR",
    },
  },
]
`;
